src/components/Viewport.ts

```diff
@@ -953,26 +953,67 @@ export default class Viewport {
     const options = this.options;
     const panelManager = this.panelManager;
 
+    const gap = options.gap;
     const viewportSize = state.size;
     const firstPanel = panelManager.firstPanel();
-    const lastPanel = panelManager.lastPanel() as Panel;
+    const lastPanel = panelManager.lastPanel()!;
 
     // There're no panels exist
     if (!firstPanel) {
       return;
     }
 
-    const sumOriginalPanelSize = lastPanel.getPosition() + lastPanel.getSize() - firstPanel.getPosition() + this.options.gap;
-    const visibleAreaSize = viewportSize + firstPanel.getRelativeAnchorPosition();
-
     // For each panels, clone itself while last panel's position + size is below viewport size
     const panels = panelManager.originalPanels();
+    const reversedPanels = panels.concat().reverse();
+
+    const sumOriginalPanelSize = lastPanel.getPosition() + lastPanel.getSize() - firstPanel.getPosition() + gap;
+    const relativeAnchorPosition = firstPanel.getRelativeAnchorPosition();
+    const relativeHangerPosition = this.getRelativeHangerPosition();
+
+    const areaPrev = (relativeHangerPosition - relativeAnchorPosition) % sumOriginalPanelSize;
+    let sizeSum = 0;
+    let panelAtLeftBoundary!: Panel;
+    for (const panel of reversedPanels) {
+      if (!panel) {
+        continue;
+      }
+      sizeSum += panel.getSize() + gap;
+      if (sizeSum >= areaPrev) {
+        panelAtLeftBoundary = panel;
+        break;
+      }
+    }
+
+    const areaNext = (viewportSize - relativeHangerPosition + relativeAnchorPosition) % sumOriginalPanelSize;
+    sizeSum = 0;
+    let panelAtRightBoundary!: Panel;
+    for (const panel of panels) {
+      if (!panel) {
+        continue;
+      }
+      sizeSum += panel.getSize() + gap;
+      if (sizeSum >= areaNext) {
+        panelAtRightBoundary = panel;
+        break;
+      }
+    }
+
+    // Need one more set of clones on prev area of original panel 0
+    const needCloneOnPrev = panelAtLeftBoundary.getIndex() <= panelAtRightBoundary.getIndex()
+      && panelAtLeftBoundary.getIndex() !== 0;
```

Comment:
@WoodNeck Although it's not important and not critical. But it would be better to check `panelAtLeftBoundary.getIndex() !== 0` first and `panelAtLeftBoundary.getIndex() <= panelAtRightBoundary.getIndex()` later. 

Because
  1. if **0 check** is failed, it doesn't need to access getIndex of Right boundary.
  2. it would be more readable if more simple condition precedes.

Reply:
That looks good, I'll fix that.